src/algorithms/split_and_merge_algorithm.cc

```diff
@@ -0,0 +1,408 @@
+      if(((z<S.size())and(k==S[z]))or((k==i)or(k==j))){
+        if(*iter_i==k){
+          if (!data_i->get_card()){ //first iteration
+            p_i+=data_i->prior_pred_lpdf(data.row(k));
+          }else{
+            p_i+=data_i->conditional_pred_lpdf(data.row(k));
+          }
+          data_i->add_datum(k, data.row(k), update_hierarchy_params());
+          iter_i++;
+        }else{
+          if (!data_j->get_card()){ //first iteration
+            p_j+=data_j->prior_pred_lpdf(data.row(k)); 
+          }
+          else{
+            p_j+=data_j->conditional_pred_lpdf(data.row(k));
+          }
+          data_j->add_datum(k, data.row(k), update_hierarchy_params());
+          iter_j++;
+        }
+        clMerge[k]=allocations[j];
+        if((k!=i)and(k!=j))z++;                                 
+      }else{
+        clMerge[k]=allocations[k];              
+      }
+      if(allocations[k]==allocations[j]){
+        if(!data_J->get_card()){
+          p_J+=data_J->prior_pred_lpdf(data.row(k));
+        }else{
+          p_J+=data_J->conditional_pred_lpdf(data.row(k));
+        }
+        data_J->add_datum(k, data.row(k), update_hierarchy_params());
+      }                                    
+    }
+    double q=1; 
+    //Fake Gibbs Sampling in order to compute the probability q
+    for(unsigned int k=0; k<S.size(); k++){
+      if(allocations_cl[k]){ //cluster j
+        cl[1]->remove_datum(S[k], data.row(S[k]), update_hierarchy_params());
+      }else{
+        cl[0]->remove_datum(S[k], data.row(S[k]), update_hierarchy_params());
+      }
+      if(cl[0]->get_card()>=1 and cl[1]->get_card()>=1){
+        double p_i=cl[0]->conditional_pred_lpdf(data.row(S[k]));
+        double p_j=cl[1]->conditional_pred_lpdf(data.row(S[k]));
+      }else{
+        if(cl[1]->get_card()==0){
+          double p_j=cl[1]->prior_pred_lpdf(data.row(S[k]));
+          double p_i=cl[0]->conditional_pred_lpdf(data.row(S[k]));
+        }else{
+           double p_i=cl[0]->prior_pred_lpdf(data.row(S[k]));
+           double p_j=cl[1]->conditional_pred_lpdf(data.row(S[k]));
+        }  
+      }
+   
+      double p=(p_i)/(p_i + p_j);
+      if(allocations[S[k]]==allocations[i]){
+        allocations_cl[k]=0;
+        cl[0]->add_datum(S[k], data.row(S[k]), update_hierarchy_params());
+        q=q*p;
+      }else{
+        allocations_cl[k]=1;
+        cl[1]->add_datum(S[k], data.row(S[k]), update_hierarchy_params());
+        q=q*(1-p);
+      }
+    }
+        
+    const double p1=q;
+    const double alpha=mixing->get_mass_new_cluster(allocations.size(), false,
+                                       true,unique_values.size());
+    // data_i->get_card()-1 = n. of data with LabelI different from i, data_j->get_card()-1= n. of data with LabelJ different from j (+1 because we are using gamma function)
+    const double p2=tgamma(data_i->get_card()-1-1+1)*tgamma(data_j->get_card()-1-1+1)/(S.size()+2-1)*alpha;
```

Comment:
Can you explain what is actually going on here? It seems to me that you're computing the prior probability of a given clustering, is it true? In that case, these numbers are specific to the DP prior, but we should be able to generalise to different mixings

Reply:
The variable p2 represents the ratio between the prior probability of the proposal cluster configuration (i.e. the split or merge configuration) and the prior probability of the current configuration. 
For the split case, the formula is:
![Cattura](https://user-images.githubusercontent.com/48724629/153051274-124da83e-bfee-4db5-93ed-7825faf0b73a.PNG)
which is used in line [183](https://github.com/vnardi/bayesmix/blob/b696efa5417915702f561fa4781dde6b7d6f1989/src/algorithms/split_and_merge_algorithm.cc#L183).
For the merge case there's a different formula w.r.t. the one we wrote in line 301, we'll check if indeed we made an error.
Anyway, the formulas are equations (3.4) and (3.5) of the paper by Jain and Neal (2004).

Given that the formula is obtained by the computation of masses of clusters, we'll work to write them in a more general way by using get_mass_new_cluster and get_mass_existing cluster.

On a side note, in the [initialization](https://github.com/vnardi/bayesmix/blob/b696efa5417915702f561fa4781dde6b7d6f1989/src/algorithms/split_and_merge_algorithm.cc#L19-L23) of the algorithm, we decided to make sure that the mixing used in the algorithm is a DP because the article by Jain and Neal studied the behavior only with that particular mixing.
Given that we are generalizing the code to work also with other mixings, should we remove this control or should we keep it until we find academic results about the correct functioning of the algorithm with a generic mixing?

Reply:
> The variable p2 represents the ratio between the prior probability of the proposal cluster configuration (i.e. the split or merge configuration) and the prior probability of the current configuration. For the split case, the formula is: ![Cattura](https://user-images.githubusercontent.com/48724629/153051274-124da83e-bfee-4db5-93ed-7825faf0b73a.PNG) which is used in line [183](https://github.com/vnardi/bayesmix/blob/b696efa5417915702f561fa4781dde6b7d6f1989/src/algorithms/split_and_merge_algorithm.cc#L183). For the merge case there's a different formula w.r.t. the one we wrote in line 301, we'll check if indeed we made an error. Anyway, the formulas are equations (3.4) and (3.5) of the paper by Jain and Neal (2004).
> 
> Given that the formula is obtained by the computation of masses of clusters, we'll work to write them in a more general way by using get_mass_new_cluster and get_mass_existing cluster.

Ok great, I think this should be taken care of in the `Mixing` object and there is no easy way to do it at the moment. Please leave a comment in the code with what exactly is this `p2` (also it should be computed in log-scale using the `lgamma` function and called `log_prior_proba`). And add
```
// FIXME@Mario: this needs generalization to other Mixings!
```
at the end of the comment block.

Reply:
The computation of `p2` (now called `log_ratio_prior_prob`) has been moved to an auxiliary function, called `compute_log_ratio_like_and_prior`. In the split case, this function populates three `Hierarchy` objects: two for the two split proposal clusters (c_i^{split} and c_j^{split}) and one for the original cluster (c_i). Before adding each datum, we update `log_ratio_prior_prob` with `get_mass_existing_cluster()`. In this way, when a DP is used, we obtain exactly the formula we reported in the last comment, but it is also possible to generalize the computation to other mixings (until this computations will be added to the `Mixing` object). 
The merge case is rather similar in the computations.

Reply:
Yes I'm actually sure that you can use split and merge with any mixing as long as you have an EPPF, several papers do so... You can remove the check of mixing being a DP!